backend/dal/Services/Concrete/ParcelService.cs

```diff
@@ -1,129 +1,158 @@
+            if (parcel.RowVersion == null) throw new RowVersionMissingException ();
+            if (this.User == null || !this.User.HasRole ("property-edit")) throw new NotAuthorizedException ();
 
-            // Only admins can update other users parcels.
-            if (!IsAllowed(entity))
-            {
-                throw new UnauthorizedAccessException();
-            }
+            var entity = this.Context.Parcels.Find (parcel.Id);
+            if (entity == null) throw new KeyNotFoundException ();
 
-            entity.Latitude = parcel.Latitude;
-            entity.Longitude = parcel.Longitude;
-            entity.UpdatedById = userId;
-            entity.UpdatedOn = DateTime.UtcNow;
-            _dbContext.SaveChanges();
-            return entity;
-        }
+            var agency_ids = this.User.GetAgencies ();
+            if (!agency_ids.Contains (entity.AgencyId)) throw new NotAuthorizedException ("User may not edit parcels outside of their agency.");
 
-        public Parcel DeleteMyParcel(Parcel parcel)
-        {
-            if (parcel == null) throw new ArgumentNullException();
-            var entity = _dbContext.Parcels.Find(parcel.Id);
-            if (entity == null) throw new KeyNotFoundException();
+            // Do not allow switching agencies through this method.
+            if (entity.AgencyId != parcel.AgencyId) throw new NotAuthorizedException ("Parcel cannot be transferred to the specified agency.");
 
-            // Only admins can update other users parcels.
-            if (!IsAllowed(entity))
-            {
-                throw new UnauthorizedAccessException();
-            }
+            this.Context.Entry (entity).CurrentValues.SetValues (parcel);
+            entity.UpdatedById = this.User.GetUserId ();
+            entity.UpdatedOn = DateTime.UtcNow;
 
-            _dbContext.Entry(entity).OriginalValues["RowVersion"] = parcel.RowVersion;
-            _dbContext.Parcels.Remove(entity);
-            _dbContext.SaveChanges();
+            this.Context.Parcels.Update (entity);
+            this.Context.CommitTransaction ();
             return entity;
         }
 
-        #region Methods
         /// <summary>
-        /// Validate that the current user is an administrator or this parcel belongs to them.
+        /// Remove the specified parcel from the datasource.
         /// </summary>
-        /// <param name="parcel">The parcel to test.</param>
-        /// <returns>True if the user is allowed.</returns>
-        public bool IsAllowed(Parcel parcel)
+        /// <param name="parcel"></param>
+        /// /// <returns></returns>
+        public void Remove (Parcel parcel)
         {
-            var userId = new Guid(this._user.FindFirstValue(ClaimTypes.NameIdentifier));
-            var isAdmin = this._user.Claims.Any(c => c.Type == ClaimTypes.Role && c.Value == "administrator");
+            if (parcel == null) throw new ArgumentNullException (nameof (parcel));
```

Comment:
At some point we could pull these three lines into a common function to make these service functions shorter, seems like only the role name changes.

Reply:
Not a bad idea.  I'll see if I can think of a reasonable place to keep it.